### ISSUE TITLE
[Bug]: Libraries getting lost on Amazon Fire devices

### ISSUE THREAD

**Issue by ghost** [nova-video-player/aos-AVP#844]
### Problem description

After initial scraping (SMBv1 share on Apple Time Capsule), both Movies and TV Shows display correctly. However, if browsing another app on the Fire device and returning to Nova Player, libraries are gone. I have to exit Nova or press 'Home', wait a couple of minutes, then launch Nova again, and the libraries will appear again. 'Recently Played' and 'Recently Added' also sometimes appear, sometimes disappear.

v6.0.67 was the last version to work correctly on FireCube 4K Gen2.
v6.0.65 and above work inconsistently on FireStick 4K Max.

Have done a complete factory reset on the FireStick 4K Max, and clean-installed Nova (v6.0.65 and above) and same problem.

### Steps to reproduce the issue

1) Setup SMBv1 share on Nova
2) Wait until initial scrape. Libraries populated correctly.
3) Play any movie or tv show.
4) Return until exit from Nova and on OS home screen.
5) Wait a few minutes or use any other app.
6) Launch Nova.

At this point, Movies and TV Shows display generic icons and no movies or tv shows appear.

7) Return until exit from Nova and on OS home screen.
8) Wait a few minutes or use any other app.
9) Launch Nova.

At this point, Movies and TV Shows display populated icons and movies and tv shows fully appear.

If I repeat the same steps again (exit, wait, launch), then libraries are gone again, etc., etc. This behavior cycles.

I have changed Initial Scan on App Startup on and off to no effect.

Cleared cache and data to no avail.

### Expected behavior

Libraries should stay populated.

### Your phone/tablet/androidTV model

Firecube 4K Gen2, Firestick 4k Max

### Operating system version

Fire OS 7.6.2.4

### Application version and app store

v6.0.97 from Amazon App Store

### Additional system information

v6.0.67 was the last version to work correctly on FireCube 4K Gen2.
v6.0.65 and above work inconsistently on FireStick 4K Max. Haven't tried earlier versions.

**Comment by courville** [comment]
It seems that your problem is not fire device specific but most probably linked to time capsule SMB1 server since it works on my side on a Synology NAS.

**Comment by Pentaphon** [comment]
I've seen this too on my Fire TV 4K stick, but it never happens on my most used device, the Onn 4K. The person using it complains that the "movies are gone" and I have to revert everything back to normal. It's happened...probably 3 times in the past year? I'm using a Windows PC as a Samba server aka my main office PC, and it provides the video for all the screens in the house.

**Comment by ghost** [comment]
The curious thing about this is that it happens on the Firecube using v6.0.97. I have reverted to every version until v6.0.67, where it does not happen (which, to my mind, discards the Time Capsule as the culprit). Same thing happens with the Firestick 4K Max, except that v6.0.67 happens sporadically on that one. I'm just guessing here, but I think it has more to do with Nova's database than the Time Capsule...

Also, this happens every single time I repeat the steps in the original post, no matter if it's 2-3 minutes apart or 1-2 hours, etc., it happens all the same.

**Comment by courville** [comment]
OK I will test. And thanks for pointing out that this would be a regression after v6.0.67. I will check all the code modifications.
Note to self 6.0.67 is "Sun May 22 16:44:03 2022 +0200"

**Comment by ghost** [comment]
Thanks. If you need logs or anything and point me out how to go about getting them, I'd be happy to help. I can also post a (private) video (on Youtube or whatever) to show you what's going on, if it helps.

**Comment by courville** [comment]
@Molasar49, is this correct to say that you do not have the problem on all your fire devices with v6.0.64?
I do not make sense why it should work with v6.0.67 on FireCube 4K Gen2 and v6.0.65 not working on FireStick 4K Max.
I will try to reproduce with the FireStick 4K Max.

**Comment by ghost** [comment]
@courville I installed v6.0.65-20220508 on Firestick 4K Max, but it showed artifacts during playback, so I uninstalled (before checking the issue posted above) and reinstalled v6.0.67-20220605. This version shows the issue posted above. I tried v6.0.71-20220623 and v6.0.97 (the one you sent me previously and also downloaded from the App Store) and they all show the same issue (on the Firestick 4K Max). I went as far as restoring to factory settings, updating the OS and ONLY installing those Nova versions, and same problem persists.

So I headed over to the FireCube 4K Gen2 and notice that only v6.0.67 works correctly - all newer versions show the same behavior. When launching on FireCube, v6.0.67 shows generic Movies and TV Shows icons but immediately recovers (split-second later), showing "customized" icons before I get a chance to do anything and all works fine.

I don't know why v6.0.67 works on FireCube but not on Firestick Max. With Firestick, I sometimes get a Recently Played movie, which plays fine, but going into 'Movies', the library is empty - which makes me think it's not an SMBv1 or Time Capsule issue, but a Nova database issue - just my (educated) guess (since I used to be a full-time .NET developer).

I was meaning to copy some of my movies from the Time Capsule to a Windows Server share and observe the behavior. Would that help in some way?

What is strange (on Firestick) is that launching shows the behavior, returning to Home (OS) and relaunching seems to make Nova recover and populate library. Returning to Home (OS) and relaunching shows the issue, etc., etc. Going back and forth populates, then empty, then populated, then empty, etc. There seems to be no scanning involved - it's just as if it reads it's database, then it doesn't, then it reads it, then it doesn't, etc. (just guessing here). There is hardly any delay. It's just there or it isn't (if that makes any sense).

**Comment by courville** [comment]
Thanks for the detailed behavior analysis. I ruled out the SMB1 timecapsule issue.
I tried to reproduce on Firestick 4K Max on latest release and SMB2 share on a synology and unfortunately it works as expected on my side. I do not have this split second of generic icon at all.

What I propose since you have the issue systematically, is to stick with latest release and try to get logs. I will share shortly a debug build with more logs around nova db and scraping. If you can capture adb logs with this release it would help (https://github.com/nova-video-player/aos-AVP/wiki/Debugging, adb connect adb logcat etc...).

**Comment by ghost** [comment]
@courville, So, you recommend I uninstall v6.0.67 and install v6.0.97 on Firestick Max and capture logs, or wait until the debug build?

**Comment by courville** [comment]
@Molasar49 please use this build: https://drive.google.com/file/d/1-HoPfztgctk7Kro1lHJt1dY-W5MxWYai/view?usp=sharing
Capture adb logcat logs (it will be verbose thus if you find a way to record the logs in a text file it would be better).
You can share the results with me via email (zip files) software at courville dot org.
Thanks in advance for your help (and patience) on this issue.

**Comment by ghost** [comment]
@courville Ok, will proceed with this right now. And on the contrary, thank you for your time and effort!

**Comment by courville** [comment]
Do not hesitate to request help.

**Comment by ghost** [comment]
@courville Wow! This version works! The only thing I notice is that it's displaying generic Movies and TV Shows icons (after successful full scan), other than that, works as expected. I captured the logs. Do you wish me to send them over?

**Comment by ghost** [comment]
Oh wait! When I disconnected adb, same problem!!!!

**Comment by ghost** [comment]
Capturing new log right now. Recently Played shows 1 movie and it plays correctly. Libraries are empty. After a few minutes, Recently Added and Recently Played are gone too, but now showing populated libraries (and generic icons). Will play around a bit and send over logs.

**Comment by courville** [comment]
Damn, yes please send the logs and I will analyze tomorrow.
Also one thing if you manage to do it is to dump nova media database before the full lib gets populated back (check "Get a dump of nova media database" section in the https://github.com/nova-video-player/aos-AVP/wiki/Debugging wiki).
If you could send me both versions (normal, and empty) via email, it would be nice too.

**Comment by ghost** [comment]
@courville Done. Also sent a link on the email to a video capture with my phone showing the issue. Hope it helps.

**Comment by courville** [comment]
Thanks for taking the time to investigate and for sharing the logs and the media dbs at different stages and an explanatory video.
I need to spend more time on the analysis.
What I can see from the media db is that all the videos are there and I see no scanning retrigered.
What I suspects now is that some loaders are stuck for some reasons that do not deliver the right information to the various fragments.
Let's dig in further.

**Comment by courville** [comment]
I will revert the changes made around the 6.0.67 and share a new build with you tonight. I will add logs only on these parts and I will kindly request that you capture logs from nova only via for linux/macos `adb shell 'logcat --pid=$(pidof -s org.courville.nova)'` and `adb shell "logcat --pid=$(pidof -s org.courville.nova)"` most probably for windows.

**Comment by ghost** [comment]
Thanks. Will be away for Sabbath but will continue after that. Thanks again!

**Comment by courville** [comment]
@Molasar49, I will share an experimental build soon for you to test. What is not entirely logic on my side looking at the code changes is that you state that it works with v6.0.67 but not with v6.0.68. Do you confirm this statement?

**Comment by courville** [comment]
OK re-reading the full thread, issue happens with v6.0.67 but less frequently but not with v6.0.65. It is thus between v6.0.65 and v6.0.67 that the issue has been introduced and it far more visible after v6.0.67.
The change I target is the throttling that I removed that could on the fire devices (for unknown reasons) to get all the loaders stuck.

**Comment by courville** [comment]
Note to self: might be related to #141, #605, #134 (check with in Video `tig  --follow ./src/main/java/com/archos/mediacenter/video/browser/loader/`)

* 19/5/27-28: introduce in VideoLoader custom `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));` #141 (4.3+?)
* 19/10/02: introduce 2s throttling in global VideoLoader to reduce E/SQLiteLog: (9) statement aborts messages #134 (5.1+)
* 21/7/11: use default VideoLoader ModernAsyncTask Executor to try to improve latency (not conclusive?) (6.0.10+)
* 22/2/15: disable 2s throttling in global VideoLoader (6.0.45+)
* 22/4/4: introduce throttling in all loaders dependent on Videoloader
* 22/4/4: get back to VideoLoader custom `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));` with 5s throttling (6.0.57+)
* 22/5/8: rework throttling infra of 5s for VideoLoaders (6.0.65+)
* 22/5/13: disable throttling for one day... and centralize throttling for loaders dependent on VideoLoader on which throttling is to be applied (6.0.65+)
* 22/5/14: enabled throttling #134 (6.0.65+)
* 22/11/12: VideoLoader LinkedBlockingQueue has increased size from 5200 to max to address #605 (6.0.82+)
* 22/11/12: MoviesSelectionLoader with threadPool of 1 is introduced to avoid blinking in VideoByFragment but penalizing the other loaders that get back to ModernAsyncTask parameters except size of LinkedBlockingQueue (6.0.82+)
* 22/11/21: VideoLoader introduce a 1h throttle delay for MoviesByLoader (6.0.86+)
* 23/1/7: VideoLoader disable throttling and give up on 1h throttle delay since causing videos not to loading in EpisodesByDate/MoviesBy(Year|Genre) (6.0.93+)

Github releases dates:
* 22/5/8: 6.0.65
* 22/5/15: 6.0.66
* 22/6/5: 6.0.67

Comment: AsyncTaskLoader relies on ModernTaskLoader now and uses `ThreadPoolExecutor(5, 128, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(10), tocheck)`.

**Comment by courville** [comment]
@Molasar49 could you please test this new build and report if it is any better https://drive.google.com/file/d/1-HoPfztgctk7Kro1lHJt1dY-W5MxWYai/view?usp=sharing ?

It has VideoLoader throttling enabled for all related loaders and relies on `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));`. However throttling is disabled in VideoSelectionLoader and it relies on default `ThreadPoolExecutor(5, 128, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(10))`.

**Comment by courville** [comment]
@Molasar49 **after and only after** testing above build, if it is still nok, it would be nice if you could provide feedback on this build where the default android `ThreadPoolExecutors` are used for the loaders (throttling is still enabled): https://drive.google.com/file/d/1-2FOkcJks9SYLNzvtZg7ludju0tuvRxN/view?usp=sharing

**Comment by ghost** [comment]
@courville On it in a couple of hours full steam and will report back. Thank you again for your efforts!

**Comment by ghost** [comment]
@courville I just installed the first version you kindly sent me (v6.1.2 2023-0319.1233) and these are my preliminary observations (on the Firestick 4K Max):

After initial scan, Movies and TV Shows libraries were populated. Returned to Home (OS), reentered and got it to fail quickly (libraries empty). After that, tried very hard to make it fail and couldn't.

I acivated the Re-scan when opening the application and got it to fail on the very next cycle (Exit-Home-Launch), but couldn't get it to fail after.

Noticed the Movie and TV Shows icons are generic no matter what.

I was just about to declare this version a winner.

Exited to Home (OS), came to write this, and before I finished, went and launched Nova just to find both libraries empty. So i guess the lengthened "pause" on Home screen (OS) and relaunching causes it to fail.

Exited to Home for 2 seconds, relaunched and now even Recently Played and Recently Added are gone, along with both libraries.

What do you want me to do next?

**Comment by ghost** [comment]
@courville After a couple of fast cycles (Exit-Home-Launch), everything populated now...

I can kind of notice a pattern on how I can get it to fail. This version is clearly harder to make fail than v6.0.67.

It's very hard to make it fail, but it does happen rarely. Generic icons, though.

A 1 to 2 minute exit to Home then relaunch seems to make it fail sometimes (clearly not always), losing Recently Played and Recently Added sometimes too.

Quick update: Now I'm having a hard time getting it to repopulate libraries.

**Comment by ghost** [comment]
Ok, just installed new version (v6.1.2 20230319.1310). Everything went smooth upon installing and initial scan. Exited to Home (OS), waited 1 minute (chronometer in hand), launched and... empty libraries :(

Recently Played and Recently Added are populated, just nothing within both libraries.

If you need logs, dbs or video, please let me know.

Update 1: Repeated the same cycle: Exit, 1 minute on Home, relaunch, now Recently Added and Recently Played are gone but libraries are populated.

Update 2: Repeated the same cycle again: Exit, Home (1 min), Relaunch. Recently Added and Recently Played are populated, both libraries are gone.

Update 3: Repeated same cycle. Recents (gone) and Libraries (populated), i.e.: flipped again.

Generic icons no matter what, in all cases.

All the above on the Firestick 4K Max. Have done no tests on the Cube 4K Gen2.

**Comment by courville** [comment]
@Molasar49 thanks for your thorough testing and reporting.

Couple of comments:
* the fact that you get generic icons in the boxes, means that somehow the custom ones are not created and it proves that some loaders are getting stuck (note to self: check which ones and put logs)
* reading your comments v6.1.2 20230319.1310 is worse than v6.1.2 2023-0319.1233 (and the later is not perfect). Comment if this is not correct.

Could you please test this build (same as v6.1.2 20230319.1310 with throttling disabled and some logs on iconBox builders):  https://drive.google.com/file/d/1EVWCbLogGMCwxIGFk38dgy3NmVIU32AX/view?usp=sharing
- `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));`
- `THROTTLE = false;`

I would be interested in the logs this time (not the db which is populated always) and let's focus on logs from `MainFragment` and `buildAllMoviesBoxTask`.

To be clear and do not induce false hopes, I do not anticipate this build to be the magic bullet.

Thx

**Comment by courville** [comment]
OK the generic icon for all movies, sagas, all tv shows is indeed a bug (regression): reverting 5e7554cb9d3c1073c3efe093d82fd811e524242a solves it.

**Comment by courville** [comment]
@Molasar49 just updated previous link https://drive.google.com/file/d/1EVWCbLogGMCwxIGFk38dgy3NmVIU32AX/view?usp=sharing with the fix for all movies icon and same config.

**Comment by courville** [comment]
Note to self inspect 9fabacac4e34646ab427043e4a4a3ae387f9b508 next for MainFragment and previous ones for cause of regression on fireTV.

**Comment by ghost** [comment]
Commencing trials with v6.1.2 (20230319.2025) on Firestick 4K Max with full log capture starting before initial scan...

Will get back soon.

**Comment by ghost** [comment]
@courville Whoa! I think we have a winner! I absolutely could not get this one to fail. Generic icons are updated correctly (with a split second delay), libraries and Recents all correct. Applied the 1-minute hypothesis, then 5 seconds, 10, 15 seconds and could not get this one to fail. I captured the full log. Sending logs now...

**Comment by courville** [comment]
OK we are finally getting somewhere (I hope).
Let's test this build to be sure where I revert back to the last release loaders behavior and implement the icon fix.
If this solves your issue, I will make a release for everyone.
Here is the new build to test (pretty please): https://drive.google.com/file/d/1EVWCbLogGMCwxIGFk38dgy3NmVIU32AX/view?usp=sharing (same link new build).

**Comment by ghost** [comment]
@courville Absolutely! Will get this fully tested later tonight and post back. Thank you for your time, effort and patience!

**Comment by ghost** [comment]
@courville Commencing trial run v6.1.1 (20230319.2230).... 

Wait...is this version right?

**Comment by ghost** [comment]
First observations:

1) Upon initial scrape (and finalized), Recently Added is populated, but Movie and TV Shows icons remain generic.
2) Both libraries are indeed populated (but icons remain generic).
3) Play any movie. Return. Recently Played correctly populated. Generic icons remain.
4) Return to Home (OS). Wait 1 minute. Relaunch. Recently Added and Recently Played populated, no more generic icons (have been changed to customized).
5) Libraries correctly populated.

Will now run different cycle tests, i.e.: Exit->Home (OS)->Relaunch and report back.

**Comment by ghost** [comment]
@courville Happy to report v6.1.1 (20230319.2230) is a clear winner on Firestick 4K Max.

Absolutely could not make this version fail no matter what I threw at it. Only caveat was the generic icons upon initial scrape finish. 1 cycle fixed it.

Commencing trial runs on Firecube 4K Gen 2. Will report back when done.

**Comment by ghost** [comment]
v6.1.1 (20230319.2230) on Firecube 4K Gen2:

1) Seeing same generic icon behavior upon initial scrape finish.
2) One cycle (Exit->Home->Relaunch) without any pause fixes the icons.
3) Probably not related. Have "Searching" and "Missing" movies; Collection shows Missing/Searching 2-Movie Collection icon, but only 1 movie there: "Missing". Other collections, i.e.: Ant-Man, Battlestar Galactica are grouping correctly. Didn't notice but same happens on Firestick 4K Max. Used Rescrape all collections option on both, still just 1 movie there. Movies are called Missing.mp4 and Searching.mp4 in their respective directories (i.e.: Media/Movies/Searching and Media/Movies/Missing).

Other than that, could not make this version fail no matter what.

Happy to report v6.1.1 (20230319.2230) is a clear winner also on Firecube 4K Gen2.

Please let me know if you need any further testing, logs, dbs, etc.

**Comment by courville** [comment]
@Molasar49 thanks for your detailed feedback and testing!
I take that the latest build works but you see a degradation of the generic icons generation that could be linked to the size of `LinkedBlockingQueue`.
I would be tempted to have you do another test on a build that I will share tonight if it is not too much to ask to address the above issue.

**Comment by ghost** [comment]
@courville Will be more than happy to test!

**Comment by courville** [comment]
@Molasar49: here is my release candidate version for you to check if it is ok on your side. If you confirm it is ok, I push it to the world ;-)
https://drive.google.com/file/d/1fRP50x5dp-9EgdLgJwK8eKbsId6eRsMo/view?usp=sharing

**Comment by ghost** [comment]
@courville Sure! Will get on it later tonight and report back. Thanks!

**Comment by ghost** [comment]
Commencing trial runs v6.1.2 (20230320.2258) on Firestick 4K Max (Wifi)...

**Comment by ghost** [comment]
@courville 

1) After initial scrape (530+ and 105+ items), Movies and TV Shows still showing generic icons.
2) Both libraries are populated correctly.
3) Entering Movies library and returning to main screen still showing generic Movie icon.
4) Entering TV Shows library and returning to main screen still showing generic TV Shows icon.
5) One cycle (Exit->Home->Relaunch) without any pause fixes both icons.

Will continue to put it thru its paces, though the initial icons issue remains...

**Comment by ghost** [comment]
@courville 

Conducted the following tests:

With "Re-scan when opening the application" set to "No":
1) Exit->Home (no pause)->Relaunch
2) Exit->Home (5s pause)->Relaunch
3) Exit->Home (10s pause)->Relaunch
4) Exit->Home (15s pause)->Relaunch
5) Exit->Home (30s pause)->Relaunch
6) Exit->Home (1m pause)->Relaunch
7) Exit->Home (3m pause)->Relaunch
8) Exit->Home (6m pause)->Relaunch

Played random movie between random cycles. Launched other random apps during longer pauses, also between random cycles.

With "Re-scan when opening the application" set to "Yes":
1) Exit->Home (no pause)->Relaunch
2) Exit->Home (5s pause)->Relaunch
3) Exit->Home (10s pause)->Relaunch
4) Exit->Home (15s pause)->Relaunch
5) Exit->Home (30s pause)->Relaunch
6) Exit->Home (1m pause)->Relaunch
7) Exit->Home (3m pause)->Relaunch
8) Exit->Home (6m pause)->Relaunch

Played random movie between random cycles. Launched other random apps during longer pauses, also between random cycles.

Everything works as expected, except for the initial scrape and generic icons not refreshed issue.

Commencing trial runs on Firecube 4K Gen2 (Wired). Will report back.

**Comment by ghost** [comment]
@courville 

After initial scrape (530+ and 105+ items), Movies and TV Shows still showing generic icons.

1) Both libraries are populated correctly.
2) Entering Movies library and returning to main screen still showing generic Movie icon.
3) Entering TV Shows library and returning to main screen still showing generic TV Shows icon.
4) One cycle (Exit->Home->Relaunch) without any pause DOES NOT fix both icons.
5) Recently added library correctly populated.
6) Both Movies and TV Shows libraries are now gone!

Will run same cycle tests and report back.

**Comment by ghost** [comment]
With "Re-scan when opening the application" set to "No":

1) Exit->Home (1m pause)->Relaunch.
Recently Added gone. Generic icons for both libraries remain. Both libraries gone. :(

2) Exit->Home (5s pause)->Relaunch.
Recently Added gone. Generic icons for both libraries remain. Both libraries gone.

3) Exit->Home (10s pause)->Relaunch.
Recently Added gone. Movies and TV Shows icons are now customized. Both libraries populated.
a) Since libraries are back, played 1 movie for a bit. Return. Recently Played now populated.

4) Exit->Home (15s pause)->Relaunch.
Recently Added and Recently played populated. Generic icons are now back. Both libraries are gone.

5) Exit->Home (30s pause)->Relaunch.
Recently Added and Recently Played gone. Generic icons remain. Both libraries are gone.

(At this point, went to confirm if I installed correct version. Preferences>About shows v6.1.2 (20230320.2258)

6) Exit->Home (1m pause)->Relaunch.
Recently Added and Recently Played gone. Generic icons remain. Both libraries are gone.

7) Exit->Home (3m pause)->Launched Netflix and browsed around->Exit->Relaunch Nova.
Recently Added and Recently Played gone. Customized icons back. Both libraries populated.

At this point, reboot (restart) Firecube. Waiting 1 minute (after Home displayed, for Cube to "settle", just in case) before recommencing cycles.

8) Relaunch.
Recently added and Recently Played populated. Customized icons present. Both libraries present.
a) Since libraries are back, played another movie for a bit. Return. Recently Played now showing 2 movies correctly.

9) Exit->Home (no pause)->Relaunch.
All correct, i.e.: Recently Added and Recently Played populated. Customized icons present. Both libraries present.

10) Exit->Home (5s pause)->Relaunch.
All correct.

11) Exit->Home (10s pause)->Relaunch.
Recently Added and Recently Played populated. Generic icons back again. Both libraries gone. :(

12) Exit->Home (15s pause)->Relaunch.
Recently Added and Recently Played gone. Customized icons back again. Both libraries populated.

13) Exit->Home (30s pause)->Relaunch.
All correct, i.e.: Recently Added and Recently Played populated. Customized icons present. Both libraries present.

14) Exit->Home (1m pause)->Relaunch.
All correct.

Have stopped all timed tests at this point.

**Comment by ghost** [comment]
@courville

v6.1.2 (20230320.2258) on:

a) Firestick 4K Max (Fire OS 7.6.2.4 PS7624/3338) Wifi works correctly, except for the initial scrape generic icons issue.
b) Firecube 4K Gen2 (Fire OS 7.6.2.4 PS7624/3337) Wired does not work correctly, i.e.: fails library, recents and icon tests.

Please let me know if you need any further info, logs, dbs, etc.

Thanks!

**Comment by ghost** [comment]
Update:

With "Re-scan when opening the application" set to "Yes":

Did a few random cycles (non-chronometered) and failed by the 3rd try (no pause on Home), i.e.: Recents gone, generic icons back, but libraries populated. Upon returning from Movies libraries, Recents populated, but generic icons remain.

A few more random cycles and failed again upon relaunch from Home OS screen (no pause).

Have stopped any further random tests at this point.

**Comment by courville** [comment]
@Molasar49 a big thank you about all your tests. Sad to see that we still have the issue with Firecube 4K Gen2 . Let me process all your inputs and start providing a clue.

**Comment by courville** [comment]
@Molasar49, it is frustrating that somehow I am not able to reproduce the issue you are having and I am sorry that I need to use you to troubleshoot this. Somehow, I am not yet fully convinced that the issue lies in the loaders. It is strange as well that we got a temporary release working and that the latest one is not.

Note that icons for video boxes are generated at the end after all the loaders are finished and the video lines are populated: they should appear correctly if you wait long enough. However if the library does not appear populated, they will not be regenerated: this is a symptom that something is wrong.

I have added more logs in the builds and I will need you to capture logcat logs **only** for nova  using this specific command for linux/macos `adb shell 'logcat --pid=$(pidof -s org.courville.nova)'` or `adb shell "logcat --pid=$(pidof -s org.courville.nova)"` most probably for windows (previous logs sent where somehow not exploitable fully unfortunately).
Note that nova needs to be running to launch the logcat (it will output previous logs of the process). Do not hesitate to ping me if you need help on this way of catching logs.

There are 3 builds to test (yes sorry it is a lot):
* [TEST1](https://drive.google.com/file/d/1bSxWAfY1cw84jkrYMVa7lpIRqXoAQJx6/view?usp=sharing) which is the same as before (i.e. current v6.1.2 and it should fail) i.e. no custom executor for `VideoLoader` and `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));` for `VideosSelectionLoader` and no throttling of loaders
* [TEST2](https://drive.google.com/file/d/1NlQ5lplDEbzP0TDk-FIBzj2FOtS94H69/view?usp=sharing) which has custom executor `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));` both for `VideoLoader` and `VideosSelectionLoader` and still no throttling of loaders
* [TEST3](https://drive.google.com/file/d/1fPQ_o3Q8j-L0To1pEXdxIHxbDYEbsRrr/view?usp=sharing) no custom executor and no throttling in `VideoLoader` and `VideosSelectionLoader`

Thanks again for all the testing and time dedicated to help on this issue.

**Comment by ghost** [comment]
@courville No problem!

I will initiate testing of these versions in a few minutes starting on Firecube and then on Firestick. I will post observations on the fly and send logcats as I go.

**Comment by ghost** [comment]
TEST1
Commencing trial runs v6.1.2 (20230321.2341) on Firecube 4K Gen2 Wired.

With "Re-scan when opening the application" set to "No":

1) Uninstalled previous version.
2) Installed TEST1 version.
3) Launched.
4) Initialed logcat capture as per above instructions.
5) Initial scraping in progress.
6) Initial scape done. Without doing anything, Recently Added populated.
7) Movies and TV Shows icons generic. Waiting 1 minute without doing anything.
8) 1m elapsed. Generic icons still displayed.
9) 2m elapsed. Still generic icons.
10) 3m elapsed. Still generic icons.
11) 4m elapsed. Still generic icons.
a) Will now enter Movies library... Empty. Return. Generic icons still displayed.
b) Will now enter TV Shows library... Empty. Return. Generic icons still displayed.
12) Will now perform cycle Exit->Home (no pause)->Relaunch...
a) Recently Added gone. Generic icons on both libraries.
b) Movies library empty.
c) TV Shows library empty.
13) Exit->Home (1m pause)->Relaunch
a) Recently Added gone.
b) Movies and TV Shows Icons customized.
c) Movies library populated.
d) TV Shows populated.
14) Played movie for a bit. Return.
a) Recently Added now populated.
b) Recently Played now populated (1 movie).
15) Exit->Home (no pause)->Relaunch
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons generic.
d) Movies library populated.
e) TV Shows library populated.
f) Both Movies and TV Shows icons still generic.
16) Exit->Home (5s pause)->Relaunch
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
17) Exit->Home (10s pause)->Relaunch
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
18) Exit->Home (15s pause)->Relaunch
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
19) Exit->Home (30s pause)->Relaunch
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
20) Exit->Home (1m pause)->Relaunch
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
21) Exit->Home (3m pause)->Relaunch
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
22) Exit->Home (6m pause)->Relaunch
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.

At this point, sending log via email.

**Comment by ghost** [comment]
TEST1
Commencing trial runs v6.1.2 (20230321.2341) on Firecube 4K Gen2 Wired.

With "Re-scan when opening the application" set to "Yes":

1) Changed above setting. Exit->Home (no pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
2) Exit->Home (5s pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
3) Exit->Home (10s pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
4) Exit->Home (15s pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
5) Exit->Home (30s pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
6) Exit->Home (1m pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
7) Exit->Home (3m pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
8) Exit->Home (6m pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Both Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.

At this point, sending log via email.

**Comment by ghost** [comment]
1) Uninstalling v6.1.2 (20230321.2341) on Firecube 4K Gen2 Wired in preparation for TEST2.
2) Installed TEST2. Launch. Preferences>About shows v6.1.2 (20230321.2347)
3) Start logcat capture.
4) Initial scape started.
a) Recently Added appears, but blank icons. Have noticed this populates correctly later on by itself.
b) Recently Added populated by itself.
5) Initial scrape finished.
a) Movies library icon generic.
b) TV Shows library icon generic.
6) Waiting...
a) 2.5m have elapsed. Screensaver kicked in (got distracted). Hit Return to exit from screensaver...
b) Movies library icon customized.
c) TV Shows library icon customized.
e) Movies library populated.
f) TV Shows populated.
7) Played movie for a bit. Return.
a) Recently Played now populated (1 movie).
8) Exit->Home (no pause)->Relaunch
a) Recently Added populated.
b) Recently Played populated.
b) Movies and TV Shows Icons customized.
c) Movies library populated.
d) TV Shows populated.
9) Exit->Home (5s pause)->Relaunch
a) All correct!
10) Exit->Home (10s pause)->Relaunch
a) All correct!
11) Exit->Home (15s pause)->Relaunch
a) All correct!
12) Exit->Home (30s pause)->Relaunch
a) All correct!
13) Exit->Home (1m pause)->Relaunch
a) All correct!
14) Exit->Home (3m pause)->Relaunch
a) All correct!
15) Exit->Home (6m pause)->Relaunch
a) All correct!

At this point, sending log via email.

**Comment by ghost** [comment]
TEST2
Commencing trial runs v6.1.2 (20230321.2347) on Firecube 4K Gen2 Wired.

With "Re-scan when opening the application" set to "Yes":

1) Changed above setting. Exit->Home (no pause)->Relaunch.
a) All correct!
2) Exit->Home (5s pause)->Relaunch
a) All correct!
3) Exit->Home (10s pause)->Relaunch
a) All correct!
4) Exit->Home (15s pause)->Relaunch
a) All correct!
5) Exit->Home (30s pause)->Relaunch
a) All correct!
6) Exit->Home (1m pause)->Relaunch
a) All correct!
7) Exit->Home (3m pause)->Relaunch
a) All correct!
8) Exit->Home (6m pause)->Relaunch
a) All correct!

Clear winner. At this point, sending log via email.

**Comment by ghost** [comment]
1) Uninstalling v6.1.2 (20230321.2347) on Firecube 4K Gen2 Wired in preparation for TEST3.
2) Disabling screensaver.
3) Installed TEST3. Launch. Preferences>About shows v6.1.2 (20230321.2344)
4) Start logcat capture.
5) Initial scape started.
a) Recently Added appears, but blank icons. Have noticed this populates correctly later on by itself.
b) Recently Added populated by itself.
6) Initial scrape finished.
a) Movies library icon generic.
b) TV Shows library icon generic.
7) Waiting...
a) 5m have lapsed. Still seeing Movies and TV Shows generic icons.
8) Entering Movies library.
a) Populated. Return.
b) Movies library still showing generic icon.
9) Entering TV Shows library.
a) Populated. Return.
b) TV Shows library still showing generic icon.
10) Exit->Home (no pause)->Relaunch
a) Recently Added populated.
b) Movies library icon customized.
c) TV Shows library icon customized.
e) Movies library populated.
f) TV Shows populated.
11) Played movie for a bit. Return.
a) Recently Played now populated (1 movie).
12) Exit->Home (no pause)->Relaunch
a) All correct!
13) Exit->Home (5s pause)->Relaunch
a) All correct!
14) Exit->Home (10s pause)->Relaunch
a) All correct!
15) Exit->Home (15s pause)->Relaunch
a) All correct!
16) Exit->Home (30s pause)->Relaunch
a) All correct!
17) Exit->Home (1m pause)->Relaunch
a) All correct!
18) Exit->Home (3m pause)->Relaunch
a) All correct!
19) At about 1.5m into the 6m pause test, adb returned "read: unexpected EOF!"
a) Sending logcat via email at this point.
20) Exit->Home (6m pause)->Relaunch
a) All correct!

**Comment by ghost** [comment]
TEST3
Commencing trial runs v6.1.2 (20230321.2344) on Firecube 4K Gen2 Wired.

With "Re-scan when opening the application" set to "Yes":

1) Changed above setting. Exit->Home (no pause)->Relaunch.
    a) All correct!
2) Exit->Home (5s pause)->Relaunch.
    a) All correct!
3) Exit->Home (10s pause)->Relaunch.
    a) All correct!
4) Exit->Home (15s pause)->Relaunch.
    a) All correct!
5) Exit->Home (30s pause)->Relaunch.
    a) All correct!
6) Exit->Home (1m pause)->Relaunch.
    a) All correct!
7) Exit->Home (3m pause)->Relaunch.
    a) All correct!
8) Exit->Home (6m pause)->Relaunch.
    a) All correct!

Seems like a winner too. At this point, sending log via email.

**Comment by ghost** [comment]
@courville Will continue to conduct same tests for all versions on Firestick 4K Max tomorrow (it's 9:10 p.m. here local time) and post back.

**Comment by courville** [comment]
OK thanks for the tremendous testing. Unfortunately the logs and not providing hints on the cause of the issue: I will check on my side how I can improve.
However what I take from your tests, is that:
- always at first app scrape starting from completely empty media library at the end the video box icons are not updated (after it is ok)
- TEST1 show the non populated library issue
- TEST2 and TEST3 are winners
This means that this is indeed a custom loader issue: using same custom `ThreadPoolExecutor` in `VideoLoader` and `VideosSelectionLoader` works but not a different one, and no custom `ThreadPoolExecutor` works too.

I need to chew on this.
It definitively helps.

**Comment by ghost** [comment]
@courville Will continue to conduct same tests for all versions on Firestick 4K Max tomorrow (it's 9:10 p.m. here local time) and post back.

**Comment by ghost** [comment]
@courville About to initiate all tests on Firestick 4K Max. Will post and send logs as I go.

**Comment by ghost** [comment]
TEST1
Commencing trial runs v6.1.2 (20230321.2341) on Firestick 4K Max Wifi.

With "Re-scan when opening the application" set to "No":

1) Uninstalled previous version.
2) Installed TEST1 version.
3) Launched.
4) Initialed logcat capture as per above instructions.
5) Screensaver disabled.
6) Initial scape started.
a) Recently Added appears, but blank icons. Have noticed this populates correctly later on by itself.
b) Recently Added populated by itself.
7) Initial scrape finished.
a) Movies library icon generic.
b) TV Shows library icon generic.
8) Waiting...
a) 3m have elapsed.
b) Movies library icon generic.
c) TV Shows library icon generic.
e) Movies library populated.
f) TV Shows populated.
9) Played movie for a bit. Return.
a) Recently Played now populated (1 movie).
10) Exit->Home (no pause)->Relaunch
a) Recently Added populated.
b) Recently Played populated.
b) Movies and TV Shows Icons generic.
c) Movies library empty.
d) TV Shows empty.

**Comment by courville** [comment]
You can do a lighter pass to confirm conclusions of above tests.

**Comment by ghost** [comment]
@courville Can you please clarify? You want me to perform swifter tests to spot pass or fail?

**Comment by ghost** [comment]
@courville If I understood corrently, have now proceeded to perform random cycles and can confirm TEST1 fails. Sending logs now.

**Comment by ghost** [comment]
Logs sent.

Proceeding to TEST2 on Firestick 4K Max Wifi...

**Comment by ghost** [comment]
Installed TEST2 on Firestick 4K Max. Started log capture and expediting test batteries to confirm PASS or FAIL.

Will report back.

**Comment by ghost** [comment]
TEST2 passes all quick tests, with and without Re-scan at startup set, except initial library icon refresh (needs 1 cycle to update). Sending logs now...

**Comment by ghost** [comment]
Logs sent.

Proceeding with TEST3 on Firestick 4K Max Wifi.

Will report back when finished.

**Comment by ghost** [comment]
@courville TEST3 passes all quick tests (with and without Re-scan at app start set), except library icons initial refresh (needs 1 cycle to update).

Sending logs now...

**Comment by ghost** [comment]
@courville Please let me know if you need any further info, dbs, logcats, etc.

**Comment by courville** [comment]
@Molasar49 thx for the testing and logcat! TEST2+TEST3 are thus ok on both firetv units. I have a viable release now and concrete evidence of where is the problem thanks to you.

**Comment by ghost** [comment]
@courville My pleasure! Please let me know if I can be of further assistance. It's no problem at all!

**Comment by courville** [comment]
See https://github.com/nova-video-player/aos-Video/commit/e04fc2b60d6db09b6482f2a47fbf689f986fe6fa

**Comment by courville** [comment]
@Molasar49, I think I made up my mind on the final tweaking (I hope). Here is the resulting pre-release [nova-TEST4.apk](https://drive.google.com/file/d/1IjyLp-0f9c1kSpGg2RiDLr-BACz8nSCq/view?usp=sharing)
If you could validate quickly if this is a winner on your side too, it would secure a final fix for your issue and ultimately close this issue.
Thx in advance for your help.

**Comment by ghost** [comment]
@courville Initiating quick tests on Firecube 4K Gen2 Wired with TEST4.

Preferences>About shows v6.1.2. (20230324.2012).

Initiating tests without capturing logs. Will report back when done...

We just had the first "mini fail" here...
Since Movies and TV Shows icons were not updating by themselves after the initial scrape, I did a standard, non-pause cycle:

1) (Return->Home (no pause)->Relaunch).
a) Recently Added gone (populated by itself during the initial scrape).
b) Movies and TV Shows  empty.

2) Did another cycle: (Return->Home (no pause)->Relaunch).
a) Recently Added gone (populated by itself during the initial scrape).
b) Movies and TV Shows populated.

3) Did another cycle: (Return->Home (no pause)->Relaunch).
a) Recently Added populated.
b) Movies and TV Shows icons customized.
c) Movies and TV Shows populated.
e) Played movie just a bit.

I could not get it to fail afterwards no matter what I tried (upto this point, non-timed). I would say the above are just tiny nuisances that are easily fixed with a cycle or two. Seems to be just a glitch.

Will now initiate quick tests on Firestick 4K Max Wifi with TEST4. Will report back when done.

**Comment by ghost** [comment]
Take that back....

4) Did another cycle: (Return->Home (random pause)->Relaunch).
a) Recently Added populated.
b) Recently Played populated.
c) Movies and TV Shows icons generic!
d) Movies library empty!
e) TV Shows library empty!
f) Activated Re-scan on app startup.

5) Return->Home (no pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.

6) Return->Home (random pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.

7) Return->Home (random pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.
f) De-activated Re-scan on app startup.

8) Return->Home (no pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.

9) Return->Home (random pause)->Relaunch.
a) All correct.

10) Return->Home (no pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Movies and TV Shows icons generic.
d) Movies library populated.
e) TV Shows library populated.

11) Return->Home (random pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.

12) Return->Home (no pause)->Relaunch.
a) Recently Added gone.
b) Recently Played gone.
c) Movies and TV Shows icons generic.
d) Movies library empty.
e) TV Shows library empty.

13) Return->Home (long-ish random pause)->Relaunch.
a) All correct.

14) Return->Home (no pause)->Relaunch.
a) Recently Added populated.
b) Recently Played populated.
c) Movies and TV Shows icons generic.
d) Movies library populated.
e) TV Shows library populated.

This is failing...

**Comment by ghost** [comment]
@courville Given the above tests, TEST4 is a fail on Firecube 4K Gen2 Wired.

If you wish me capture logs, dbs, etc., and/or conduct tests (TEST4) on Firestick 4K Max, will be happy to do so Sunday (since Sabbath is within the hour for us).

I really thought and hoped this was a done deal... Sad to see this fail :(

I'm here to help.

**Comment by courville** [comment]
@Molasar49 thanks again for all the testing and yes the result is very unsettling indeed :-( This means that my interpretation is not correct. 

Let's sum it up:

* [TEST1](https://drive.google.com/file/d/1bSxWAfY1cw84jkrYMVa7lpIRqXoAQJx6/view?usp=sharing) which is the same as before (i.e. current v6.1.2 and it should fail) i.e. no custom executor for `VideoLoader` and `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));` for `VideosSelectionLoader` and no throttling of loaders -> **FAILURE**
* [TEST2](https://drive.google.com/file/d/1NlQ5lplDEbzP0TDk-FIBzj2FOtS94H69/view?usp=sharing) which has custom executor `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200));` both for `VideoLoader` and `VideosSelectionLoader` and still no throttling of loaders -> **SUCCESS**
* [TEST3](https://drive.google.com/file/d/1fPQ_o3Q8j-L0To1pEXdxIHxbDYEbsRrr/view?usp=sharing) no custom executor and no throttling in `VideoLoader` and `VideosSelectionLoader` -> **SUCCESS**
* [TEST4](https://drive.google.com/file/d/1IjyLp-0f9c1kSpGg2RiDLr-BACz8nSCq/view?usp=sharing) which has identical custom executor `ThreadPoolExecutor(1, 4, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(256));` both for `VideoLoader` and `VideosSelectionLoader` and no throttling -> **FAILURE**

This contradicts my hypothesis of having identical executors for `VideoLoader` and `VideosSelectionLoader` solving the issue (which is the case of TEST2 working and TEST4 not working). However not messing with executors works too (TEST3).
I am also focusing on `VideoLoader` and `VideosSelectionLoader` while in the firetv tests, no use of `VideosSelectionLoader` is made.  This would hint that only `VideoLoader` should be considered (to be confirmed).

Thus what I need to do is:

- test of focussing on same executor for `VideoLoader` and `VideosSelectionLoader` but not customizing the `VideosSelectionLoader` since it should inherit from `VideoLoader` one, works
- revert to TEST2 config for release if it fails

At least we have configurations working (unexplained): I will share one new build today following above points if this is ok with you.

**Comment by courville** [comment]
@Molasar49, OK new round of test builds (you can skip testing TEST6 if TEST5 works): all builds have no throttling, no custom executor for `VideosSelectionLoader` all inherited from `VideoLoader`:
- [TEST5](https://drive.google.com/file/d/19n41Pb2IQZl6fqFk_teP_CXTig_M9obN/view?usp=sharing) where `VideoLoader` uses custom `ThreadPoolExecutor(1, 4, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(256))`
- [TEST6](https://drive.google.com/file/d/1G_3CA9L-4qqXqadqlSU27OEOOLkxOCtp/view?usp=sharing) where `VideoLoader` uses custom `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200))` (should be same as working TEST2 thus working but with only one customization in `VideoLoader`)
- [TEST7](https://drive.google.com/file/d/1oPX9oRJUFTd74OTa0EuAKd73I3yKOOPK/view?usp=sharing) where `VideoLoader` does not use any custom executor (i.e. it thus uses mother `ModernAsyncTask` one which is `ThreadPoolExecutor(5, 128, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(10))` on Android but no clue on fireOS)

I am sorry to impose on you all this testing but I am sure that you are not alone in seeing the issue.

**Comment by ghost** [comment]
@courville Not a problem. Will start testing tomorrow. Thanks again for all the effort!

**Comment by ghost** [comment]
@courville, Initiating TEST5 random quick tests on Firecube 4K Gen2 Wired. Will report (control and unusual observations) as I go...

**Comment by ghost** [comment]
TEST5

1) Uninstalled previous version.
2) Reboot (to provide for a clean test).
3) Wait 2 minutes for device to "settle" (finish loading any background processes).
4) Install. All permissions granted (as usual).
5) Preferences>About shows v6.1.3 (20230325.1810)
6) Initial scan in progress...
a) Recently Added populated during initial scan.
7) Waited 2 minutes and counting...
a) Movies and TV Shows icons generic.
b) Movies library populated
c) TV Shows library populated
d) Upon return from both libraries, icons still generic.
8) 1 cycle (Return->Home (no pause)->Relaunch.
a) Movies and TV Shows customized icons.
b) Movies library populated
c) TV Shows populated.
9) Played random movie.
a) Recently Played populated. 

Initiating random cycles...

**Comment by ghost** [comment]
10) Return->Home (no pause)->Relaunch.
a) Movies and TV Shows icons generic
b) Everything else correct.
11) Return->Home (2s pause)->Relaunch.
a) Movies and TV Shows icons generic
b) Recently Added populated
c) Recently Played populated
d) Movies library empty!
e) TV Shows empty!
12) Return->Home (random pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows icons customized
d) Movies library populated
e) TV Shows populated

**Comment by ghost** [comment]
It's getting harder and harder to get this one to fail, but I just did again after some random cycles:

13) Return->Home (random pause)->Relaunch.
a) Recently Added populated
b) Recently Played populated
c) Movies and TV Shows icons generic
d) Movies library empty!
e) TV Shows empty!
14) Return->Home (no pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows icons generic
d) Movies library empty!
e) TV Shows empty!

At this point (just made it fail -aka: empty libraries- 2 times in a row with random cycles), will declare this version a fail, and will continue with TEST6 in the spirit of getting all versions tested.

Can go back later and reinstall this version if you need logs, dbs etc.

**Comment by ghost** [comment]
TEST6

1) Uninstalled previous version.
2) Reboot (to provide for a clean test).
3) Wait 2 minutes for device to "settle" (finish loading any background processes).
4) Install. All permissions granted (as usual).
5) Preferences>About shows v6.1.3 (20230325.1822)
6) Initial scan in progress...
a) Recently Added populated during initial scan.
7) Waited 2 minutes and counting...
a) Movies and TV Shows icons generic.
b) Movies library populated
c) TV Shows library populated
d) Upon return from both libraries, icons still generic.
8) 1 cycle (Return->Home (no pause)->Relaunch.
a) Movies and TV Shows customized icons.
b) Movies library populated
c) TV Shows populated.
9) Played random movie.
a) Recently Played populated.

Initiating random cycles...

**Comment by ghost** [comment]
10) Return->Home (no pause)->Relaunch.
a) Movies and TV Shows icons generic
b) Everything else correct.
11) Return->Home (random pause)->Relaunch.
a) Recently Added gone!
b) Recently Played populated
c) Movies and TV Shows icons generic
d) Movies library empty!
e) TV Shows empty!
12) Return->Home (short random pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows icons customized
d) Movies library populated
e) TV Shows populated
13) Return->Home (no pause)->Relaunch.
a) Recently Added populated
b) Recently Played populated
c) Movies and TV Shows icons generic
d) Movies library gone!
e) TV Shows gone!
14) Return->Home (no pause)->Relaunch.
a) Recently Added populated
b) Recently Played populated
c) Movies and TV Shows icons generic
d) Movies library populated
e) TV Shows populated

**Comment by ghost** [comment]
Up to this point, I can sometimes get the generic icons to appear, but am having a hard time getting the libraries to appear empty...

To recap, could initially get this version to fail, but afterwards no matter what I tried, the best I could do was get the generic library icons to appear. Everything else seems correct. Big improvement over the last few versions, but from a normal user's point of view, I still think TEST3 is the best one and still reigning champion. TEST6 is a close 2nd place.

Will proceed now with TEST7.

Update: Just got TEST6 to fail again (empty libraries after about a 5 minute pause). Another random cycle and both Recents gone. TEST6 is now a distant 2nd place to TEST3.

**Comment by ghost** [comment]
TEST7

1) Uninstalled previous version.
2) Reboot (to provide for a clean test).
3) Wait 2 minutes for device to "settle" (finish loading any background processes).
4) Install. All permissions granted (as usual).
5) Preferences>About shows v6.1.3 (20230325.1833)
6) Initial scan in progress...
a) Recently Added populated during initial scan.
7) Waited 3+ minutes and counting...
a) Movies and TV Shows icons generic.
b) Movies library populated
c) TV Shows library populated
d) Upon return from both libraries, icons still generic.
8) 1 cycle (Return->Home (no pause)->Relaunch.
a) Movies and TV Shows still generic icons.
b) Movies library populated
c) TV Shows populated.
9) Played random movie.
a) Recently Played populated.
b) Still Movies and TV Shows generic icons
10) A few random cycles after...
a) Movies and TV Shows customized icons

Initiating random cycles...

**Comment by courville** [comment]
Thanks for all the testing: thinking about it and carefully reading your reports, I am now getting convinced that something is not handled properly in the `MainFragment` livecycle when exiting and re-entering and perhaps not with the `VideoLoader`...
You are insisting on the "Movies and TV Shows generic icons" a lot and it is indeed wrong. I will wait for your final results on TEST7 and will get back on analyzing `MainFragment`.

**Comment by ghost** [comment]
@courville With TEST7, can sometimes get generic icons displayed, but libraries are always populated. Could not get it to fail. TEST7 is VERY close to TEST3 (which I regard as the benchmark; just my opinion).

I can almost see a pattern getting generic icons to show... (quick non-pause cycles usually gets generic icons); other than that, seems to be VERY close being a winner. As a note, I use TEST3 as my day-to-day version for my personal viewing with my family, and have yet to make it fail (not trying to on a daily basis; just a normal use experience). To my memory, just got it to display generic icons 1 single time (could have been just a random device glitch).

Hope all this testing works. If you need logs, dbs or anything else (testing on Firestick, etc.), please let me know.

**Comment by ghost** [comment]
UPDATE: Absolutely could not get TEST7 to show empty libraries, but can get it to show generic icons with quick, non-pause cycles on (an almost consistent) basis.

Hope this helps!

**Comment by courville** [comment]
It does help. I need to process all this now. FYI TEST7 is almost TEST3.
Just a quick question: do you confirm that on all the tests above you do not have automatic rescan at application launch?

**Comment by ghost** [comment]
> It does help. I need to process all this now. FYI TEST7 is almost TEST3. Just a quick question: do you confirm that on all the tests above you do not have automatic rescan at application launch?

Yes, I did not have Re-scan on app start set to Yes. Do you want me to set this option to Yes and retest?

**Comment by courville** [comment]
Not now, let's focus on this first use case (which is the most common one).

**Comment by ghost** [comment]
@courville Got it. Anything else, let me know. Thanks!

**Comment by courville** [comment]
@Molasar49, I am able to reproduce the issue now in the emulator with a little perseverance.
I would value if you could let me know what was the worst release you tested (i.e. the one that was making it fail quite consistently)?

**Comment by ghost** [comment]
@courville I would have to say that TEST5 was the worst, followed by TEST6. TEST7 the best of this last batch. TEST3 reigns king of them all. However, they all suffer from generic icons at the end of initial scan.

TEST5 was quite easy to make fail. TEST6 was initially easy to get it to fail, then I had a hard time, but managed to get it to fail after trying hard. TEST7 extremely hard to make fail, but I could manage to get the generic icons consistently on this one.

TEST5 very unstable... could not predict failure pattern. TEST6 unstable too, but a bit more predictable. TEST7 is very resilient. TEST3 still beats it, though.

**Comment by courville** [comment]
@Molasar49 another question, to get back to the Android home screen, do you use the "back" key or the "home" key of the remote (I assume "back")?

Couple of notes:

- rereading the code, this is normal to have "generic icons" at end of the first library scan but not at any subsequent launch of the app
- I found some crashes quickly circling through nova/Android home screen due to asyncTask not finished creating the icon boxes and nova activity not running (I will share a build to see if the issue is related)
- I am unfortunately once again not able to reproduce the issue with TEST5 (for now)

**Comment by ghost** [comment]
@courville, always using "back" key on the remote on tests.

**Comment by courville** [comment]
Another day another test build: here is [TEST8](https://drive.google.com/file/d/1twLs4QLZXY23yrPC_1q2fKLMKeaDV-J0/view?usp=sharing) that is identical to TEST5 but with a fix for a crash in `MainFragment` happening cycling through Android home/Nova main activity when exiting nova and the creation of icon box was still ongoing. I have no clue if it helps but I could not reproduce on my side the "empty library issue" with this one (this is not a positive sign since I have hard time to reproduce it anyway).

I will try to reproduce more extensively tonight on a device with larger video library.

**Comment by ghost** [comment]
@courville Initiating TEST8 (quick) tests on Firecube 4K Gen2 Wired... will report as I go.

**Comment by ghost** [comment]
TEST8

1) Uninstalled previous version.
2) Reboot (to provide for a clean test).
3) Wait 2 minutes for device to "settle" (finish loading any background processes).
4) Install. All permissions granted (as usual).
5) Preferences>About shows v6.1.3 (20230327.1852)
6) Re-scan on app startup is set to "No" (default).
7) Initial scan in progress...
a) Recently Added populated during initial scan.
8) I understand Movies and TV Shows icons generic is standard, expected behavior...
a) Will wait for 2 minutes when done just as an app "settling period".
9) Initial scan finished. 2 minute "settling period" elapsed.
a) Movies and TV Shows icons generic.
b) Movies library populated
c) TV Shows library populated
d) Upon return from both libraries, icons still generic.
10) 1 cycle (Return->Home (no pause)->Relaunch.
a) All correct.
11) Played random movie.
a) Recently Played populated.
b) All correct.

Initiating random cycles... (all cycles using "Return" and "Select/Enter" key on remote

**Comment by ghost** [comment]
12) After a few random cycles, including a quick Return->Home (quite quick no pause)->Relaunch (aka flash cycle, unnatural in normal day-to-day operation).
a) Movies and TV Shows icons generic. Just a nuisance, all else correct.

Quick note: Apple Time Capsule "sleeps" after about 5 minutes of no activity. It takes about 8s+ after it "sleeps" for a request to return data. This might provide you with an insight (however, on "good" versions like TEST3, this doesn't seem to be a problem). That is why some of the cycles include 6m+ pauses.

**Comment by ghost** [comment]
Can confirm: With "flash cycles" (totally unnatural), I can consistently get generic library icons. However, all else is correct. Just my opinion this is not a flaw, as a "flash cycle" is nothing short of unnatural. Even a near 0.5s cycle causes no problems (i.e.: restores icons to customized).

**Comment by courville** [comment]
Thus it means that somehow my last change makes the worst build (TEST5) stable for the 'library not populated' issue.

For the generic icon issue: I will work on it on the next step. For now I made the app recover when crashing during buildAllMoviesBoxTask and I have a design issue regarding icons not generated after first scrap.

Anyhow, this would be a fantastic progress if you do not get anymore the 'library not populated' issue.

**Comment by ghost** [comment]
Up to this point, no matter what I threw at it (including flash cycles), I could NOT get it to fail (i.e.: empty libraries)...

Will continue a bit more cycles and then proceed to set "Re-scan on app start" to Yes, and continue tests. If it succeeds, will test TEST8 on Firestick 4K Max.

**Comment by courville** [comment]
If it works, it means that in reality the global issue is a huge race condition which frequency of occurence is influenced by the loaders strategy but not linked to the loaders itself. I will thus revert the original strategy for better efficiency.

Note that I can reproduce the generic box icons: easier then to fix.

Thanks again for all the effort in thorough testing!

**Comment by ghost** [comment]
Happy to report TEST8 is a winner on Firecube 4K Gen2 Wired with Re-scan on app startup set to "No". Will now set to "Yes" and continue tests...

**Comment by ghost** [comment]
Extremely happy to report TEST8 is a WINNER on Firecube 4K Gen2 Wired! Could not make it lose libraries no matter what I threw at it (including using Home key instead of Return, randomly). Works as expected.

Will run TEST8 trials on Firestick 4K Max Wifi after lunch and report back.

**Comment by courville** [comment]
[TEST9](https://drive.google.com/file/d/1FPShYT6gI3ru03g5PBKmU6AXtzjPEIKk/view?usp=sharing) should now always build iconBoxes except for initial first scraping at first install (otherwise it is the same as TEST8).

**Comment by ghost** [comment]
@courville About to initiate tests (TEST8) on Firestick. Do you wish me to continue, or jump right into TEST9 (on Firecube, etc.)?

**Comment by courville** [comment]
@Molasar49 just finish TEST8 (pretty please) and then you can take TEST9 for your everyday use.

**Comment by ghost** [comment]
@courville Gotcha. Initiating now...

**Comment by ghost** [comment]
TEST8 (on Firestick 4K Max Wifi)

Note: Will assume generic library icons on initial scan finish is correct behavior and should regenerate on first cycle.

1) Uninstalled previous version.
2) Reboot (to provide for a clean test).
3) Wait 2 minutes for device to "settle" (finish loading any background processes).
4) Install. All permissions granted (as usual).
5) Preferences>About shows v6.1.3 (20230327.1852)
6) Re-scan on app startup is set to "No" (default).
7) Initial scan in progress...
a) Recently Added populated during initial scan.
8) Initial scan finished. 2 minute "settling period" elapsed.
a) All correct.
9) 1 cycle (Return->Home (1s pause)->Relaunch.
a) Recently Added populated.
b) Movies and TV Shows icons generic.
c) Movies libraries gone!!!
10) Return->Home (3s pause)->Relaunch.
a) Recently Added gone!!!
b) Movies and TV Shows icons customized.
c) Movies libraries populated.

Initiating random cycles... (all cycles using "Return" and "Select/Enter" key on remote). Firestick is 12ft. aprox away from AP, clear line of sight. AP is wired to unmanaged switch, wired to managed switch, wired to Time Capsule, all 1Gb (ethernet wired) speed.

**Comment by ghost** [comment]
Just failed again...

11) Return->Home (random medium pause)->Relaunch.
a) Recently Added populated
b) Had played a random movie before cycle. Recently Played gone!
c) Movies and TV Shows icons generic.
d) Movies library empty!
e) TV Shows library empty!
12) Return->Home (random 5s-ish pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows icons generic!
d) Movies library empty!
e) TV Shows library empty!

**Comment by ghost** [comment]
Just for the sake of completeness, checked Network stats (OS). Shows:

Frequency: 5220MHz
Channel: 44
Signal strength: High (-41 dBm)
Noise: Low (-90 dBm)
Signal to Noise Ratio: High (49)

Ran integrated OS speed test. Shows:
Internet Speed (91.60 Mbps) - I have a 200Mbps up/down symmetrical link.

Continuing with tests...

**Comment by ghost** [comment]
13) Return->Home (random pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows icons generic!
d) Movies library empty!
e) TV Shows library empty!

Since Recently Added is populated, picked a random title, and it played correctly. However, on return, Recently Played is still gone...

**Comment by ghost** [comment]
14) Return->Home (random pause)->Relaunch.
a) Everything is gone!

@courville This is clearly failing (easily) on Firestick 4K Max. What do you wish me to do?

**Comment by ghost** [comment]
15) Return->Home (random medium pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows icons generic!
d) Movies library empty!
e) TV Shows library empty!

Rebooting AP and managed switch for the sake of completeness and continue test. Will report back.

**Comment by ghost** [comment]
Rebooted Firestick last. Cycles in progress... looks promising...

Whoops... just failed again :(

Also, sometimes still losing Recently Added and Recently Played.

Just hit another fail (empty libraries) after a random 8s-ish pause.

Yet another fail (empty libraries) after random (20s-ish) cycle...

TEST8 is definitely a no-go on Firestick 4K Max Wifi... :(

After another random cycle, libraries icons customized and both libraries populated, but both Recents gone, which (to my mind at least) should rule out connectivity issues. Any movie plays fine. When both libraries are empty but Recents are populated, any title from Recents also plays fine. Would I be correct to assume this (not a connectivity or SMB issue)?

**Comment by ghost** [comment]
Again, for the sake of completeness, uninstalled TEST8 and installed TEST3 (without rebooting) - Firestick 4K Max.

After 1 hour of intense, UNRELENTING POUNDING (well beyond the scope of normal use), I did manage to make it fail about 4 times, but it was VERY hard to do so...

Going to extremes, have relocated Firestick to a location where I can use the ethernet adapter (included) and wire it to the same switch where the Firecube is connected. Wifi is now turned OFF. Using a computer monitor (HP) as test screen.

Will reinitiate TEST8 tests and report back...

**Comment by ghost** [comment]
TEST8 (on Firestick 4K Max Wired)

Note: Will assume generic library icons on initial scan finish is correct behavior and should regenerate on first cycle.

1) Uninstalled previous version.
2) Reboot (to provide for a clean test).
3) Wait 2 minutes for device to "settle" (finish loading any background processes).
4) Install. All permissions granted (as usual).
5) Preferences>About shows v6.1.3 (20230327.1852)
6) Re-scan on app startup is set to "No" (default).
7) Initial scan in progress...
a) Recently Added populated during initial scan.
8) Initial scan finished. 2 minute "settling period" elapsed.
a) All correct.
9) 1 cycle (Return->Home (1s pause)->Relaunch.
a) All correct.
10) Played random movie.
a) Recently Played populated.
b) All correct.
11) Return->Home (random short pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows icons generic!
d) Movies library empty!
e) TV Shows library empty!
12) Return->Home (random short pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows icons customized.
d) Movies library populated.
e) TV Shows library populated.
13) Return->Home (random medium pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows icons generic!
d) Movies library empty!
e) TV Shows library empty!

Can definitely say without any further tests that the above results pretty much discard wifi issues as the Firestick is now ethernet wired to the same (unmanaged) 100Mbps switch as the Firecube 4K Gen2.

**Comment by ghost** [comment]
14) Return->Home (random medium pause)->Relaunch.
a) Everything gone!
b) Apple Time Capsule shortcut visible in Network->Indexed folders.
c) Opened shortcut and all media visible (mp4 and/or mkv files).
d) Upon return, everything is still missing (i.e.: Recently Added / Played gone, generic icons, libraries empty).
15) Return->Home (random medium pause)->Relaunch.
a) Everything gone!
16) Return->Home (random short pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows library icons customized.
d) Movies library populated.
e) TV Shows library populated.
17) Return->Home (random long-ish pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows library icons generic.
d) Movies library empty!
e) TV Shows library empty!

@courville I think this shows (sufficient) erratic behavior on Firestick 4K Max Wired (discarding wifi issues) to be declared a definite fail for TEST8. I have no idea why the difference in behavior between Cube and Stick...

Please let me know if you wish any other tests, etc.

**Comment by ghost** [comment]
Switched to another port on the Dell ethernet switch (as I just saw the SPD/LNK/ACT LED -maybe- a bit dimmed on the previous port).

Both SPD/LNK/ACT (blinking) and FDX/HDX LEDs fully bright. Continuing tests...

18) Return->Home (random long-ish pause)->Relaunch.
a) All correct.
b) Played another random movie.
c) All correct.

**Comment by ghost** [comment]
Well, that was short-lived.

Just got it to fail after a random pause.

19) Return->Home (random short-ish pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows library icons generic.
d) Movies library empty!
e) TV Shows library empty!
20) Return->Home (random short-ish pause)->Relaunch.
a) Recently Added gone!
b) Recently Played gone!
c) Movies and TV Shows library icons customized.
d) Movies library populated
e) TV Shows library populated
20) Return->Home (random short-ish pause)->Relaunch.
a) All correct.
21) Return->Home (random medium-ish pause)->Relaunch.
a) Recently Added populated
b) Recently Played gone!
c) Movies and TV Shows library icons generic!
d) Movies library gone!
e) TV Shows library gone!
22) Return->Home (random short-ish pause)->Relaunch.
a) Everything gone!
23) Return->Home (random long-ish pause)->Relaunch.
a) Everything gone!

I think we can safely assume the observed failures are not wifi related (or even connectivity related), nor even Apple Time Capsule (sleeping) related, as TEST8 does not fail on Firecube.

I didn't POUND it (TEST8) relentlessly on the Cube, but I can if you want to. Please let me know. As a note, these latter (wired) tests weren't merciless either; about the same as on the Cube. I am now moving the Stick back to its original place (wifi).

**Comment by courville** [comment]
Hi, thanks again for all the tests, even if this is disappointing. Thus conclusion is that all the releases are failing ultimately even the best of all TEST3 :-(
At this point, I  think it always failed somehow but hardly noticeable (nothing that a back and relaunch could fix).
Perhaps this is linked to AsyncTask being deprecated https://developer.android.com/reference/android/os/AsyncTask for a good reason:

> AsyncTask was intended to enable proper and easy use of the UI thread. However, the most common use case was for integrating into UI, and that would cause Context leaks, missed callbacks, or crashes on configuration changes. It also has inconsistent behavior on different versions of the platform, swallows exceptions from doInBackground, and does not provide much utility over using [Executor](https://developer.android.com/reference/java/util/concurrent/Executor)s directly.

At this point, I would be tempted to comply and try out the proposed `java.util.concurrent` or Kotlin concurrency utilities (livedata etc.) but this is far more work.

**Comment by ghost** [comment]
@courville Yes, I reinstalled TEST3 on Firestick (while it was still wired) and I made it fail quite frequently with just a bit of trying... :(

I don't know why it's that much more noticeable on the Stick than on the Cube... Maybe because the hardware is more capable or faster on the Cube... no idea...

**Comment by courville** [comment]
@Molasar49 thanks for the confirmation of my hunch. I guess the difference of behavior is linked to a core race condition that is indeed put in evidence depending on the hardware capabilities and patterns.

**Comment by courville** [comment]
Since this is not a loader issue: get safely back to `ThreadPoolExecutor(1, 1, 10, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(5200))` since using a smaller queue makes application crash while scanning large video collection entering the `MoviesByYearFragment` for instance:
```
E/AndroidRuntime(17769): java.util.concurrent.RejectedExecutionException: Task androidx.loader.content.ModernAsyncTask$3@da21126 rejected from java.util.concurrent.ThreadPoolExecutor@a523ef2[Running, pool size = 4, active threads = 4, queued tasks = 256, completed tasks = 360]
E/AndroidRuntime(17769): 	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2086)
E/AndroidRuntime(17769): 	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:848)
E/AndroidRuntime(17769): 	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1394)
E/AndroidRuntime(17769): 	at androidx.loader.content.ModernAsyncTask.executeOnExecutor(ModernAsyncTask.java:452)
E/AndroidRuntime(17769): 	at androidx.loader.content.AsyncTaskLoader.executePendingTask(AsyncTaskLoader.java:228)
E/AndroidRuntime(17769): 	at androidx.loader.content.AsyncTaskLoader.dispatchOnCancelled(AsyncTaskLoader.java:241)
E/AndroidRuntime(17769): 	at androidx.loader.content.AsyncTaskLoader$LoadTask.onCancelled(AsyncTaskLoader.java:94)
E/AndroidRuntime(17769): 	at androidx.loader.content.ModernAsyncTask.finish(ModernAsyncTask.java:488)
E/AndroidRuntime(17769): 	at androidx.loader.content.ModernAsyncTask$InternalHandler.handleMessage(ModernAsyncTask.java:507)
E/AndroidRuntime(17769): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime(17769): 	at android.os.Looper.loop(Looper.java:223)
E/AndroidRuntime(17769): 	at android.app.ActivityThread.main(ActivityThread.java:7668)
E/AndroidRuntime(17769): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(17769): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/AndroidRuntime(17769): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```
Conclusion: one could benefit of multithread but since a long `LinkedBlockingQueue` is required not to get a crash this will not help because multithreads kicks in when queue is full...

See https://github.com/nova-video-player/aos-Video/commit/c9636f0317e958e96cbeaf740aea634c84bb21f4

**Comment by Pentaphon** [comment]
@Molasar49 has this issue been resolved for you?

**Comment by ghost** [comment]
Have done no tests on any newer versions as it was my understanding that this wasn't getting fixed any time soon (i.e., there was no indication to test out any newer versions). I am still on version v6.1.3 (20230327.1852). Have not updated to any newer version. I had a sense that the issue would be addressed much later (since it appears to  involve a lot of rework). Has anything changed regarding this issue in later versions? Will be happy to test.
